src/datatrove/pipeline/stats/base.py

```diff
@@ -0,0 +1,112 @@
+            Each group in top_k_groups will truncate the statistics to the top k keys.
+            This lowers memory usage and speeds up the merging in second-stage.
+    """
+
+    type = "📊 - STATS"
+    name = "👑 Summary stats"
+    _requires_dependencies = ["tldextract"]
+
+    def __init__(
+        self,
+        output_folder: DataFolderLike,
+        groups_to_compute: list[GROUP] | None = None,
+        histogram_round_digits: int = 3,
+        top_k_config: TopKConfig = DEFAULT_TOP_K_CONFIG,
+    ) -> None:
+        from tldextract import TLDExtract
+
+        super().__init__()
+        self.output_folder = get_datafolder(output_folder)
+        self.groups = groups_to_compute or list(get_args(GROUP))
+        self.histogram_round_digits = histogram_round_digits
+        self.top_k_cfg = top_k_config
+        self.tld_extractor = TLDExtract()
+
+    @abstractmethod
+    def extract_stats(self, doc: Document) -> dict[str, int | float]:
+        """
+        Abstract method for extracting stats from a document.
+        Args:
+            doc: The document to extract stats from.
+
+        Returns:
+            A dictionary of statistics, where the key is the stat name and the value is the stat value.
+        """
+        raise NotImplementedError()
+
+    def get_kv(self, doc: Document, value: STAT_TYPE, group_name: GROUP) -> tuple[str, STAT_TYPE]:
+        if group_name == "histogram":
+            # Use rounding to reduce then number of values for histogram
+            return str(round(value, self.histogram_round_digits)), 1
+        elif group_name == "summary":
+            return "summary", value
+        elif group_name == "fqdn":
+            return doc.metadata.setdefault("fqdn", self.tld_extractor.extract_str(doc.metadata["url"]).fqdn), value
+        elif group_name == "suffix":
+            return doc.metadata.setdefault("suffix", self.tld_extractor.extract_str(doc.metadata["url"]).suffix), value
+        else:
+            raise ValueError(f"Unknown group name: {group_name}")
+
+    def run(self, data: DocumentsPipeline, rank: int = 0, world_size: int = 1) -> DocumentsPipeline:
+        groups_dicts: dict[GROUP, dict[str, MetricStatsDict]] = {
+            group: defaultdict(MetricStatsDict) for group in self.groups
+        }
+        with self.track_time():
+            for doc in data:
+                try:
+                    doc_stats = self.extract_stats(doc)
+                except Exception as e:
+                    logger.error(f"Error while extracting stats from document {doc.id}", exc_info=e)
+                    raise e
+
+                for group, counters in groups_dicts.items():
+                    for stat, value in doc_stats.items():
+                        key, value = self.get_kv(doc, value, group)
+                        counters[stat][key] += value
+
+                doc.metadata.update(doc_stats)
+                yield doc
```

Comment:
this should be outside the block with track time (put the track time inside the loop)

Reply:
It's still inside the track_time xD
```suggestion
            yield doc
```